content/posts/2019-08-03---Going-from-ember-to-vue.md

```diff
@@ -0,0 +1,284 @@
+})
+```
+
+To this in Vue:
+
+```javascript
+fullName() {
+  return `${this.firstName} ${this.lastName}`;
+}
+```
+
+Computed properties have the same issues in Vue as Ember, in that it won't observe nested keys more than one level deep and it has some issues with observing arrays.
+
+## The Router
+
+![A Router](https://i.imgur.com/xlUWWN7.jpg?1)
+<aside>Photo by <a href="https://www.pexels.com/@rawpixel">Rawpixel on Pexels</a></aside>
+
+The router in Vue feels like a good mix between react-router and the Ember Router.
+Let's just look at a full example of a router in Vue:
+
+```javascript
+import Vue from 'vue';
+import Router from 'vue-router';
+import { auth } from 'utils';
+
+import { NotFound, Dashboard, Login } from 'routes';
+
+Vue.use(Router);
+
+const router = new Router({
+  mode: 'history',
+  routes: [
+    {
+      path: '*',
+      component: NotFound
+    },
+    {
+      path: '/dashboard',
+      name: 'Dashboard',
+      meta: {
+        layout: 'sidebar',
+        requiresAuth: true
+      },
+      component: Dashboard
+    },
+    {
+      path: '/login',
+      name: 'Login',
+      component: Login,
+      // If they were redirected here, don't bother doing our auth check
+      beforeEnter: async (to, from, next) => {
+        if (to.params.redirected) {
+          next();
+        } else {
+          if (auth.currentUser) {
+            next({ name: 'Dashboard' });
+          } else {
+            next();
+          }
+        }
+      }
+    }
+  ]
+});
+
+router.beforeEach(async (to, from, next) => {
+  const { currentUser } = auth;
+
+  // If the page requires authentication
+  // make sure they are logged in
+  // if they aren't, kick them to the login page
+  if (to.meta.requiresAuth) {
+    if (currentUser) {
+      next();
+    } else {
+      next({
+        name: 'Login',
+        params: { redirected: true } // indicate they are redirected
+      });
+    }
+  } else {
+    next(); // business as usual
+  }
+});
+
+export default router;
+```
+
+Much like react-router, we specify the route name and point it to a component that we have imported (here, we store them in a `routes/index.js` file, but they could live anywhere).
+We can specify a path, just like in Ember but we can also attach additional information in any format we want.
+This becomes really handy for when you want to mark some pages as public and some that need authentication.
+In Ember, you would do that in the `route.js` file for a particular route/path but in Vue, you handle it all here.
+This can be really nice because it allows us to either write one large `beforEach` function, like you see at the bottom of the file or you can write those functions individually for every route (like you see for the `login` route).
+
+So, the pattern is very similar to routes in Ember, but you are keeping your logic in one file rather than spread across a ton of different route object files that may or may not extend from one another.
```

Comment:
route files in ember are also optional.

also, there is typically not much duplication between routes in ember

Reply:
I can definitely clarify this. What I meant is that if you want specific logic before entering a route (like hooking into `beforeModel` in a route object), you can handle it for all routes in one place, rather than each `route.js` file or using something like `ember-simple-auth`'s `AuthenticatedRoute` mixin.

Reply:
you only need to handle that at the parent route though -- which is places like Heroku have a `protected` route, and that one route has the auth logic, and then all sub-routes kinda just natively have auth checking, because an ancestor route does

Reply:
@sbatson5 it is almost the same exact code in Ember: https://github.com/ember-learn/guides-source/blob/master/app/router.js#L17-L19

Reply:
> you only need to handle that at the parent route though -- which is places like Heroku have a protected route, and that one route has the auth logic, and then all sub-routes kinda just natively have auth checking, because an ancestor route does

But then you need to put all your child routes inside that `protected` folder.